### PR TITLE
Add subscribe notices for products without subscription

### PR DESCRIPTION
**Pull request by vishnugopal**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Fixes https://github.com/Automattic/woocommerce.com/issues/21455

Add a new notice to the plugins list for products used without subscriptions reminding them to purchase.

CleanShot 2024-09-04 at 09 45 56@2x

[As described here](https://arc.net/l/quote/viyavoce), this also removes notices for expired and expiring subscriptions from WooCommerce settings screens.

Add a new notice to the WooCommerce screens for missing subscriptions:

CleanShot 2024-09-04 at 09 45 27@2x
CleanShot 2024-09-02 at 10 25 44@2x

This is part of the PU2 project, [project thread](https://woomarketplace.wordpress.com/2024/06/20/project-thread-prevent-usage-without-active-subscription-phase-2/).

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Connect your WC Core site to woocommerce.test [as described in this guide](https://wp.me/p6JqRr-cDo).
2. Make sure you have some purchased subscriptions in woocommerce.test.
3. Install a plugin without an active subscription, and make sure that you see the notices above in the plugins list and the WooCommerce screens.
4. Make sure that the notices for expired and expiring subscriptions remain unchanged.
CleanShot 2024-08-30 at 15 01 49@2x
5. The notices for missing subscriptions in WooCommerce screens should only show up if you don't have any expired or expiring subscriptions. i.e. missing subscriptions have the lowest priority.
CleanShot 2024-09-02 at 10 36 37@2x
7. Make sure that expiring and expired subscriptions are not visible in the settings screens., i.e. this should not appear:
CleanShot 2024-09-02 at 10 39 58@2x
8. For missing and expired subscriptions, the link at both the plugin screen and extensions screen should point to the add to cart page. Check this both with a single product and multiple products. Click on the link and ensure that products are added to cart:
CleanShot 2024-09-11 at 11 27 21@2x
CleanShot 2024-09-11 at 11 28 33@2x
8. Please test for any edge cases.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add a new notice to the plugins list for products used without subscriptions reminding them to purchase.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php [hunk 1]
```diff
@@ -340,6 +352,108 @@ function ( $subscription ) {
...
+	}
+
+	/**
+	 * Gets the price of a product without a subscription as that is not present in the update data.
+	 * This uses the WooCommerce.com API to get the price of the product.
+	 *
+	 * @param mixed $product_id The product ID.
+	 * @return null|string The price of the product.
+	 */
+	public static function get_regular_price_for_product_without_subscription( $product_id ) {
```
Comment (mehamasum) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: @vishnugopal Since the price is not available directly, and we need to make an extra API call, does it make sense to get rid of the price info from this alert? So the notices would be without the ` for %s`.  

We were hoping to minimize the number of external calls (cc: @gedex)

Comment (vishnugopal) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: @mehamasum  I can do that as well if that's the design decision we're taking. I have tried to follow best practices and added caching and timeouts for this API call, but will defer to your guidance here.

also cc @poligilad-auto

Comment (gedex) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: I agree with removing extra remote request just for the price info.

Comment (vishnugopal) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: Removed in da6854304f. I've also updated screenshots.

File: plugins/woocommerce/src/Admin/PluginsHelper.php [hunk 2]
```diff
@@ -970,6 +982,86 @@ function ( $sub ) {
...
+			'missing',
+		);
+
+		$button_link = add_query_arg(
+			array(
+				'utm_source'   => 'pu',
+				'utm_campaign' => 'pu_in_apps_screen_purchase',
+			),
+			self::WOO_SUBSCRIPTION_PAGE_URL
+		);
```
Comment (mehamasum) [inline plugins/woocommerce/src/Admin/PluginsHelper.php]: This should be the [same "subscribe" link used in-app](https://github.com/woocommerce/woocommerce/blob/904dfaa2ffe948d191010dbf74d799abbaf99875/plugins/woocommerce-admin/client/marketplace/utils/functions.tsx#L455-L459). Sending folks to wccom subscription page can be confusing, & doesn't help them subscribe to a product.

Comment (vishnugopal) [reply plugins/woocommerce/src/Admin/PluginsHelper.php]: Thanks @mehamasum changed in https://github.com/woocommerce/woocommerce/pull/51060/commits/72b43acb050d346d7716b82ac91628e3ebed8971, I also changed the link for expired subscriptions in https://github.com/woocommerce/woocommerce/pull/51060/commits/e22f170a04973fcb240ed6f2622d7f1531f6f9a6 and left expiring subscriptions alone as the CTA for that is auto-renew.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php [hunk 3]
```diff
@@ -340,6 +353,56 @@ function ( $subscription ) {
...
+		// Extract product ID from the response.
+		$product_id = preg_replace( '/[^0-9]/', '', $response->id );
+
+		wc_get_logger()->debug( 'Checking if plugin has subscription', array( 'product_id' => $product_id ) );
+
+		if ( WC_Helper::has_product_subscription( $product_id ) ) {
+			return;
+		}
+
+		wc_get_logger()->debug( 'Nope' );
```
Comment (hesahesa) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: is this a test logger?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php [hunk 4]
```diff
@@ -340,6 +353,56 @@ function ( $subscription ) {
...
+	 * @param object $plugin_data An array of plugin metadata.
+	 * @param object $response  An object of metadata about the available plugin update.
+	 *
+	 * @return void.
+	 */
+	public static function display_notice_for_plugins_without_subscription( $plugin_data, $response ) {
+		// Extract product ID from the response.
+		$product_id = preg_replace( '/[^0-9]/', '', $response->id );
+
+		wc_get_logger()->debug( 'Checking if plugin has subscription', array( 'product_id' => $product_id ) );
```
Comment (hesahesa) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: also this one, is this a test logger?

File: plugins/woocommerce/src/Admin/PluginsHelper.php [hunk 5]
```diff
@@ -37,13 +37,18 @@ class PluginsHelper {
 	 *
 	 * @var bool
 	 */
-	public static $can_show_expiring_subs_notice = true;
+	public static $can_show_expiring_or_missing_subs_notice = true;
```
Comment (hesahesa) [inline plugins/woocommerce/src/Admin/PluginsHelper.php]: this one variable serves as a flag whether a notice is already shown / not (so the other notices should not be shown)

back when we only have 2 notices the naming makes sense, however since currently we have 3 types of notices, maybe it's time to rename this? maybe something like `subscription_notice_already_shown`?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php (hunk 4 above)
Comment (vishnugopal) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: Thank you, fixed in https://github.com/woocommerce/woocommerce/pull/51060/commits/a6c781bd25c3d0eac054cf8d2765d5cffaff570d

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php (hunk 3 above)
Comment (vishnugopal) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-updater.php]: Thank you, fixed in https://github.com/woocommerce/woocommerce/pull/51060/commits/a6c781bd25c3d0eac054cf8d2765d5cffaff570d

File: plugins/woocommerce/src/Admin/PluginsHelper.php (hunk 2 above)
Comment (vishnugopal) [reply plugins/woocommerce/src/Admin/PluginsHelper.php]: I've also updated test cases above to reflect this change.

**Review by hesahesa — CHANGES_REQUESTED** [review]
have some feedbacks, would be good to address that

File: plugins/woocommerce/src/Admin/PluginsHelper.php (hunk 5 above)
Comment (vishnugopal) [reply plugins/woocommerce/src/Admin/PluginsHelper.php]: Thanks, this makes a lot of sense. I renamed this in 0e9b92fecf. I also added a docbloc to better describe this as there is a clear priority of notification order.

Comment (hesahesa) [reply plugins/woocommerce/src/Admin/PluginsHelper.php]: also if we use this name: `subscription_notice_already_shown` i realized that the boolean logic needs to be flipped. Feel free to use other variable name of flip the logic accordingly

Comment (vishnugopal) [reply plugins/woocommerce/src/Admin/PluginsHelper.php]: Ah, thanks again, flipped in https://github.com/woocommerce/woocommerce/pull/51060/commits/8df6d945e4eadd6e328624caf3a0645fead0d110

**Comment by hesahesa** [comment]
image

image

tested the changes in this PR

**Review by hesahesa — APPROVED** [review]
LGTM, thank you @vishnugopal
